File: tests/contracts/workspace-image-release.test.ts
import assert from "node:assert/strict";
import test from "node:test";

import {
  decodeWorkspaceImageReleaseCatalog,
  type WorkspaceImageReleaseCatalog
} from "../../packages/contracts/workspace-image-release.ts";

const installedImage = `registry.example.com/opl/workspace@sha256:${"b".repeat(64)}`;
const rollbackImage = `registry.example.com/opl/workspace@sha256:${"a".repeat(64)}`;

test("Workspace image release catalog accepts unique immutable versions containing the installed image", () => {
  const input: WorkspaceImageReleaseCatalog = {
    schemaVersion: 1,
    releases: [
      { version: "26.8.26", image: installedImage },
      { version: "26.8.4", image: rollbackImage }
    ]
  };
  assert.deepEqual(decodeWorkspaceImageReleaseCatalog(JSON.stringify(input), installedImage), input);
});

test("Workspace image release catalog rejects mutable, duplicate, unknown, and incomplete inputs", () => {
  for (const input of [
    { schemaVersion: 1, releases: [{ version: "latest", image: "registry.example.com/opl/workspace:latest" }] },
    { schemaVersion: 1, releases: [{ version: "same", image: installedImage }, { version: "same", image: rollbackImage }] },
    { schemaVersion: 1, releases: [{ version: "26.8.4", image: rollbackImage }] },
    { schemaVersion: 1, releases: [{ version: "26.8.26", image: installedImage, mutable: true }] }
  ]) {
    assert.equal(decodeWorkspaceImageReleaseCatalog(JSON.stringify(input), installedImage), null);
  }
});

File: packages/contracts/go/workspace_image_release_test.go
package contracts

import (
	"strings"
	"testing"
)

func workspaceImageForTest(digit string) string {
	return "registry.example/opl/workspace@sha256:" + strings.Repeat(digit, 64)
}

func TestDecodeWorkspaceImageReleaseCatalogFallsBackToInstalledImage(t *testing.T) {
	installed := workspaceImageForTest("a")
	catalog, err := DecodeWorkspaceImageReleaseCatalog("", installed)
	if err != nil || catalog.SchemaVersion != 1 || len(catalog.Releases) != 1 || catalog.Releases[0].Version != "installed" || catalog.Releases[0].Image != installed {
		t.Fatalf("catalog=%#v err=%v", catalog, err)
	}
}

func TestDecodeWorkspaceImageReleaseCatalogAcceptsApprovedImmutableVersions(t *testing.T) {
	stable, rollback := workspaceImageForTest("b"), workspaceImageForTest("a")
	raw := `{"schemaVersion":1,"releases":[{"version":"26.8.26","image":"` + stable + `"},{"version":"26.8.4","image":"` + rollback + `"}]}`
	catalog, err := DecodeWorkspaceImageReleaseCatalog(raw, stable)
	if err != nil || !catalog.ContainsImage(stable) || !catalog.ContainsImage(rollback) {
		t.Fatalf("catalog=%#v err=%v", catalog, err)
	}
	if release, found := catalog.ReleaseForImage(rollback); !found || release.Version != "26.8.4" || WorkspaceImageDigest(release.Image) == "" {
		t.Fatalf("rollback release=%#v found=%v", release, found)
	}
}

func TestDecodeWorkspaceImageReleaseCatalogRejectsUnprotectedShapes(t *testing.T) {
	installed, other := workspaceImageForTest("b"), workspaceImageForTest("a")
	for name, raw := range map[string]string{
		"tag":               `{"schemaVersion":1,"releases":[{"version":"latest","image":"registry.example/opl/workspace:latest"}]}`,
		"missing installed": `{"schemaVersion":1,"releases":[{"version":"old","image":"` + other + `"}]}`,
		"duplicate version": `{"schemaVersion":1,"releases":[{"version":"same","image":"` + installed + `"},{"version":"same","image":"` + other + `"}]}`,
		"unknown field":     `{"schemaVersion":1,"releases":[{"version":"current","image":"` + installed + `","mutable":true}]}`,
	} {
		t.Run(name, func(t *testing.T) {
			if catalog, err := DecodeWorkspaceImageReleaseCatalog(raw, installed); err == nil {
				t.Fatalf("catalog unexpectedly accepted: %#v", catalog)
			}
		})
	}
}
